infra/gcp/bash/ensure-terraform.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+
+# Copyright 2022 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script ensures terraform service accounts for prow jobs and GCB are created.
+# DOCS: https://cloud.google.com/build/docs/securing-builds/configure-user-specified-service-accounts#cross-project_set_up
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_DIR=$(dirname "${BASH_SOURCE[0]}")
+. "${SCRIPT_DIR}/lib.sh"
+
+function usage() {
+    echo "usage: $0" > /dev/stderr
+    echo > /dev/stderr
+}
+
+if [ $# != 0 ]; then
+    usage
+    exit 1
+fi
+
+# The GCP project name.
+PROJECT=$(k8s_infra_project "prow" "k8s-infra-terraform")
+
+color 6 "Ensuring project exists: ${PROJECT}"
+ensure_project "${PROJECT}"
+
+PROJECT_SERVICES=(
+    iam.googleapis.com
+)
+
+# Enable GCB APIs
+color 6 "Ensure required services are enabled for: ${PROJECT}"
+
+ensure_only_services "${PROJECT}" "${PROJECT_SERVICES[@]}"
+
+# Create various service accounts for Terraform
+color 6 "Creating oci-proxy dev Service Account"
+ensure_service_account \
+    "${PROJECT}" \
+    "oci-proxy-dev" \
+    "oci proxy dev Terraform Account"
+
+color 6 "Creating oci-proxy prod Service Account"
+ensure_service_account \
+    "${PROJECT}" \
+    "oci-proxy-prod" \
+    "oci proxy prod Terraform Account"
+
+oci_proxy_dev_account="$(svc_acct_email "${project}" "${STAGING_SIGNER_SVCACCT}")"
+color 6 "Ensuring GCB service agent of k8s-staging-infra-tools can impersonate ${oci_proxy_dev_account} "
+ensure_serviceaccount_role_binding \
+    "${oci_proxy_dev_account}" \
+    "serviceAccount:service-1017132094926@gcp-sa-cloudbuild.iam.gserviceaccount.com" \
+    "roles/iam.serviceAccountTokenCreator"
+
+oci_proxy_prod_account="$(svc_acct_email "${project}" "${STAGING_SIGNER_SVCACCT}")"
+color 6 "Ensuring GCB service agent of k8s-staging-infra-tools can impersonate ${oci_proxy_prod_account} "
```

Comment:
We want to deploy using a promoted image of `archeio`. why do we need to use the staging project ?

Reply:
The GCB job that will run terraform will be in the `k8s-staging-infra-tools` project. Since it runs in that project, it needs to impersonate the designated service accounts in the `k8s-infra-terraform` project.

Reply:
Why not simply use a postsubmit to do that ? we could simply have a service account associated to do that.

Reply:
That is the plan, but GCB is running a different account.

https://github.com/kubernetes-sigs/oci-proxy/blob/main/cloudbuild.yaml is running as `1017132094926@cloudbuild.gserviceaccount.com` which isn't what we want as we need 2 different service accounts used for separate jobs. Unlike most jobs, we need to run terraform plan in presubmits which is a privileged operation, hence the need for multiple service accounts.

We could grant Owner to `1017132094926@cloudbuild.gserviceaccount.com` on both projects but I don't like that idea.